src/tools/rbd_mirror/Replayer.cc

```diff
@@ -220,6 +224,115 @@ int Replayer::init()
+  PoolImageIds images;
+  list<pair<int64_t, string> > pools;
+  int r = m_local->pool_list2(pools);
+  if (r < 0) {
+    derr << "error listing pools: " << cpp_strerror(r) << dendl;
+    return;
+  }
+
+  for (auto kv : pools) {
+    int64_t pool_id = kv.first;
+    string pool_name = kv.second;
+    int64_t base_tier;
+    r = m_local->pool_get_base_tier(pool_id, &base_tier);
+    if (r == -ENOENT) {
+      dout(10) << "pool " << pool_name << " no longer exists" << dendl;
+      continue;
+    } else if (r < 0) {
+      derr << "Error retrieving base tier for pool " << pool_name << dendl;
+      continue;
+    }
+    if (pool_id != base_tier) {
+      // pool is a cache; skip it
+      continue;
+    }
+
+    librados::IoCtx ioctx;
+    r = m_local->ioctx_create2(pool_id, ioctx);
+    if (r == -ENOENT) {
+      dout(10) << "pool " << pool_name << " no longer exists" << dendl;
+      continue;
+    } else if (r < 0) {
+      derr << "Error accessing pool " << pool_name << cpp_strerror(r) << dendl;
+      continue;
+    }
+
+    rbd_mirror_mode_t mirror_mode;
+    r = librbd::mirror_mode_get(ioctx, &mirror_mode);
+    if (r < 0) {
+      derr << "could not tell whether mirroring was enabled for " << pool_name
+	   << " : " << cpp_strerror(r) << dendl;
+      continue;
+    }
+    if (mirror_mode == RBD_MIRROR_MODE_DISABLED) {
+      dout(20) << "pool " << pool_name << " has mirroring disabled" << dendl;
+      continue;
+    }
+
+    librados::IoCtx remote_ioctx;
+    r = m_remote->ioctx_create(ioctx.get_pool_name().c_str(), remote_ioctx);
+    if (r < 0 && r != -ENOENT) {
+      dout(10) << "Error connecting to remote pool " << ioctx.get_pool_name()
+               << ": " << cpp_strerror(r) << dendl;
+      continue;
+    } else if (r == -ENOENT) {
+      // TODO: schedule local images for deletion!
+      continue;
+    }
+
+    pool_id = remote_ioctx.get_id();
+
+    std::set<ImageIds> image_ids;
+    std::string last_read = "";
+    int max_read = 1024;
+    do {
+      std::map<std::string, std::string> mirror_images;
+      r = librbd::cls_client::mirror_image_list(&ioctx, last_read, max_read,
+                                                &mirror_images);
+      if (r < 0) {
+        derr << "error listing mirrored image directory: "
+             << cpp_strerror(r) << dendl;
+        continue;
+      }
+      for (auto it = mirror_images.begin(); it != mirror_images.end(); ++it) {
+        std::string remote_image_id;
+        r = librbd::cls_client::mirror_image_get_image_id(&remote_ioctx,
+                                                          it->second,
+                                                          &remote_image_id);
+        if (r == -ENOENT) {
+          dout(20) << "remote image with global id " << it->second << "does "
+                   << "not exist. Scheduling delete of local mirror" << dendl;
+          m_image_deleter->schedule_image_delete(ioctx.get_id(), it->first,
```

Comment:
Very important to not schedule any images for deletion if they are locally flagged as primary.  Need to guard here and also on shut-down (unless both deletion paths were combined in `set_sources` as noted above).